File: R/trex_selector_methods/trex/demo_trex_05_mc_sim_dummy_distributions/demo_trex_05_mc_sim_dummy_distributions.R
# ==============================================================================
# demo_trex_05_mc_sim_dummy_distributions.R
# ==============================================================================
#
# Classical T-Rex selector demo: a Monte Carlo simulation comparing the dummy
# distributions available for the T-Rex Selector across three base solvers
# (TLARS, TOMP, TAFS), with the STANDARD L-loop strategy held fixed.
#
# Base solvers compared (one result file pair per solver):
#   TLARS — equiangular LARS path (stagnation stop AUTO-resolves to disabled).
#   TOMP  — greedy orthogonal matching pursuit (stagnation stop enabled).
#   TAFS  — greedy adaptive forward selection, rho_afs = 1.0 (stagnation stop
#           enabled).
#
# Mirrors:
# cpp/trex_selector_methods/trex/demo_trex_05_mc_sim_dummy_distributions.cpp
#
# The T-Rex FDR calibration rests on exchangeability between real null
# predictors and the injected dummy variables. Since dummies are centered and
# L2-normalized before entering the solver, their scale is immaterial — this
# demo probes whether the SHAPE of the dummy distribution (tails, discreteness,
# skewness, sparsity, cross-column dependence) affects the achieved FDR / TPR
# and the calibrated dummy multiplier L / stopping time T.
#
# Demo content (see trex_dummy_distribution()):
# ----------------------------------------------------------------
#
#    Normal              — N(0, 1); the reference choice.
#    Uniform             — U(-sqrt(3), sqrt(3)); bounded, light tails.
#    Rademacher          — {-1, +1} equiprobable; discrete two-point.
#    StudentT (df 3 / 5) — heavy tails (unit-variance scaled for df > 2).
#    Laplace             — double exponential; heavier-than-normal tails.
#    Gumbel              — extreme-value; skewed (centered at 0).
#    Triangle            — bounded, symmetric; (-sqrt(6), 0, sqrt(6)).
#    Logistic            — between normal and Laplace tails.
#    Mammen              — asymmetric golden-ratio two-point distribution.
#    SparseRademacher    — constrained sparse Rademacher, sparsity s = 0.1.
#    UniformSphere       — uniform on the unit 5-sphere; consecutive 5-column
#                          groups are dependent (norm constraint), a deliberate
#                          probe of the exchangeability axis. The library
#                          requests dummies in multiples of p, so dim must
#                          divide p (the default dim = 3 would throw at
#                          p = 1000, hence dim = 5).
#
#  Scenario:
#  ----------
#  High-dimensional  (n = 300, p = 1000, s = 10).
#  SNR sweep: {0.1, 0.2, ..., 2.0, 5.0}  (21 values).
#
#  Two support scenarios:
#    random support  — redrawn per trial;
#    block support   — contiguous block {1, 2, ..., s}.
#  Reports averaged FDR / TPR / Avg L / Avg T per distribution x SNR.
#
# ==============================================================================

library(TRexSelectorNeo)
library(parallel)

# ==============================================================================
# Setup
# ==============================================================================

this_dir_ <- tryCatch(
  dirname(normalizePath(sys.frame(1L)$ofile)),
  error = function(e) {
    args <- commandArgs(trailingOnly = FALSE)
    file_arg <- grep("--file=", args, value = TRUE)
    if (length(file_arg) > 0) dirname(normalizePath(sub("--file=", "", file_arg[1]))) else "."
  }
)

source(file.path(this_dir_, "..", "..", "support_generators.R"))
source(file.path(this_dir_, "..", "..", "simulation_utils.R"))
source(file.path(this_dir_, "..", "trex_sim_utils.R"))


# ==============================================================================
# Global Simulation Parameters
# ==============================================================================

OUT_DIR <- file.path(this_dir_, "simulation_results", "data")
dir.create(OUT_DIR, recursive = TRUE, showWarnings = FALSE)

num_cores <- 6L
NUM_MC <- 10L
run_trex_05 <- TRUE

# ==============================================================================
# Base solvers to compare (outer sweep; rho_afs only used by TAFS)
# ==============================================================================

SOLVERS <- list(
  list(name = "TLARS", solver = "TLARS", rho_afs = NULL),
  list(name = "TOMP",  solver = "TOMP",  rho_afs = NULL),
  list(name = "TAFS",  solver = "TAFS",  rho_afs = 1.0)
)


# ==============================================================================
# Dummy distributions to compare
# ==============================================================================

# name : row label used in output tables and CSV.
# dist : list built by trex_dummy_distribution(), passed to
#        trex_control(dummy_distribution = ...).

.make_dist_entry <- function(name, dist) {
  list(name = name, dist = dist)
}

DISTRIBUTIONS <- list(
  .make_dist_entry("Normal",         trex_dummy_distribution("Normal")),
  .make_dist_entry("Uniform",        trex_dummy_distribution("Uniform")),
  .make_dist_entry("Rademacher",     trex_dummy_distribution("Rademacher")),
  .make_dist_entry("StudentT_df3",   trex_dummy_distribution("StudentT", df = 3)),
  .make_dist_entry("StudentT_df5",   trex_dummy_distribution("StudentT", df = 5)),
  .make_dist_entry("Laplace",        trex_dummy_distribution("Laplace")),
  .make_dist_entry("Gumbel",         trex_dummy_distribution("Gumbel")),
  .make_dist_entry("Triangle",       trex_dummy_distribution("Triangle")),
  .make_dist_entry("Logistic",       trex_dummy_distribution("Logistic")),
  .make_dist_entry("Mammen",         trex_dummy_distribution("Mammen")),
  .make_dist_entry("SparseRad_s0.1",
                   trex_dummy_distribution("ConstrainedSparseRademacher", s = 0.1)),
  .make_dist_entry("UnifSphere_d5",
                   trex_dummy_distribution("UniformSphere", dim = 5L))
)


# ==============================================================================
# MC simulation — variable true support, dummy distribution comparison
# ==============================================================================

trx_05_dummy_distributions <- function(num_MC = NUM_MC, rnd_coef = FALSE,
                                       block_support = FALSE) {

  n    <- 300L
  p    <- 1000L
  s    <- 10L
  tFDR <- 0.10
  snr_values <- c(seq(0.1, 2.0, by = 0.1), 5.0)

  cat("\n", strrep("=", 70L), "\n", sep = "")
  cat("MC Simulation \xe2\x80\x94 Dummy Distribution Comparison  |  TLARS\n")
  cat("  High-dimensional (p > n)\n")
  cat(sprintf("  n = %d,  p = %d,  s = %d,  num_MC = %d  [%s support]\n\n",
              n, p, s, num_MC, if (block_support) "block" else "random"))

  n_dist <- length(DISTRIBUTIONS)
  n_snr  <- length(snr_values)
  dist_names <- vapply(DISTRIBUTIONS, `[[`, "", "name")

  # Solver sweep — one result file pair (txt + csv) per solver
  for (solver_entry in SOLVERS) {

  # Result containers: distribution x SNR (reset per solver)
  fdr_mat   <- matrix(0.0, nrow = n_dist, ncol = n_snr)
  tpr_mat   <- matrix(0.0, nrow = n_dist, ncol = n_snr)
  avg_L_mat <- matrix(0.0, nrow = n_dist, ncol = n_snr)
  avg_T_mat <- matrix(0.0, nrow = n_dist, ncol = n_snr)
  rownames(fdr_mat) <- rownames(tpr_mat) <-
    rownames(avg_L_mat) <- rownames(avg_T_mat) <- dist_names
  colnames(fdr_mat) <- colnames(tpr_mat) <-
    colnames(avg_L_mat) <- colnames(avg_T_mat) <- as.character(snr_values)


  # Simulation sweep: iterate over dummy distributions and SNR values
  for (di in seq_along(DISTRIBUTIONS)) {
    dist_entry <- DISTRIBUTIONS[[di]]

    ctrl_args <- list(
      solver               = solver_entry$solver,
      K                    = 20L,
      max_dummy_multiplier = 10L,
      use_max_T_stop       = TRUE,
      lloop_strategy       = "STANDARD",
      dummy_distribution   = dist_entry$dist
    )
    if (!is.null(solver_entry$rho_afs)) ctrl_args$rho_afs <- solver_entry$rho_afs
    ctrl <- do.call(TRexSelectorNeo::trex_control, ctrl_args)

    cat(sprintf("%s\nSolver: %s  |  Dummy distribution: %s\n%s\n\n",
                strrep("=", 50L), solver_entry$name, dist_entry$name,
                strrep("=", 50L)))

    for (snr_idx in seq_along(snr_values)) {
      snr       <- snr_values[[snr_idx]]
      base_seed <- 24L + (snr_idx - 1L) * 1000L

      make_data <- local({
        current_snr <- snr
        use_block   <- block_support
        function(mc, trial_seed) {
          trial_support <- if (use_block) seq_len(s) else
            make_support_random(s, p, seed = trial_seed)
          trial_coefs <- rep(1.0, s)
          dgp_gauss_snr(n, p, trial_support, coefs = trial_coefs,
                        snr = current_snr, seed = trial_seed)
        }
      })

      lbl    <- sprintf("SNR=%.2f [%s/%s]", snr, solver_entry$name, dist_entry$name)
      result <- .run_mc_trex(make_data, ctrl, tFDR, base_seed, num_MC, num_cores,
                             track_L_T = TRUE, label = lbl)

      fdr_mat[di, snr_idx]   <- result$mean_FDR
      tpr_mat[di, snr_idx]   <- result$mean_TPR
      avg_L_mat[di, snr_idx] <- result$mean_L
      avg_T_mat[di, snr_idx] <- result$mean_T
    }
    cat("\n")
  }

  stem <- sprintf("demo_trex_05_dummy_distributions_results_n%d_p%d_%s_%s",
                  n, p, tolower(solver_entry$name),
                  if (block_support) "block_support" else "random_support")
  .save_and_print_mc(OUT_DIR, stem, num_MC, dist_names, snr_values,
                     fdr_mat, tpr_mat, avg_L_mat, avg_T_mat)
  cat("\n")

  }  # end solver sweep
}


# ==============================================================================
# Main
# --------------
# Run demo 05
# ==============================================================================
if (run_trex_05) {
  trx_05_dummy_distributions(NUM_MC, rnd_coef = FALSE, block_support = FALSE)
  # trx_05_dummy_distributions(NUM_MC, rnd_coef = FALSE, block_support = TRUE)
}
cat("\ndemo_trex_05_mc_sim_dummy_distributions.R complete.\n")

File: Python/trex_selector_methods/trex/demo_trex_05_mc_sim_dummy_distributions/demo_trex_05_mc_sim_dummy_distributions.py
# ==============================================================================
# demo_trex_05_mc_sim_dummy_distributions.py
# ==============================================================================
#
# Classical T-Rex selector demo — dummy distribution comparison across three
# base solvers (TLARS, TOMP, TAFS), with the STANDARD L-loop strategy fixed.
# Mirrors R/trex_selector_methods/trex/demo_trex_05_mc_sim_dummy_distributions.R and
# cpp/trex_selector_methods/trex/demo_trex_05_mc_sim_dummy_distributions.cpp.
#
# Base solvers compared (one result file pair per solver):
#   TLARS — equiangular LARS path (stagnation stop AUTO-resolves to disabled).
#   TOMP  — greedy orthogonal matching pursuit (stagnation stop AUTO-resolves
#           to enabled).
#   TAFS  — greedy adaptive forward selection, rho_afs = 0.3 (stagnation stop
#           AUTO-resolves to enabled).
#
# The T-Rex FDR calibration rests on exchangeability between real null
# predictors and the injected dummy variables. Since dummies are centered and
# L2-normalized before entering the solver, their scale is immaterial — this
# demo probes whether the SHAPE of the dummy distribution (tails, discreteness,
# skewness, sparsity, cross-column dependence) affects the achieved FDR / TPR
# and the calibrated dummy multiplier L / stopping time T.
#
# Demo content (spec dicts resolved by trex_sim_common._make_dummy_distribution):
#
#    Normal              — N(0, 1); the reference choice.
#    Uniform             — U(-sqrt(3), sqrt(3)); bounded, light tails.
#    Rademacher          — {-1, +1} equiprobable; discrete two-point.
#    StudentT (df 3 / 5) — heavy tails (unit-variance scaled for df > 2).
#    Laplace             — double exponential; heavier-than-normal tails.
#    Gumbel              — extreme-value; skewed (centered at 0).
#    Triangle            — bounded, symmetric; (-sqrt(6), 0, sqrt(6)).
#    Logistic            — between normal and Laplace tails.
#    Mammen              — asymmetric golden-ratio two-point distribution.
#    SparseRademacher    — constrained sparse Rademacher, sparsity s = 0.1.
#    UniformSphere       — uniform on the unit 5-sphere; consecutive 5-column
#                          groups are dependent (norm constraint), a deliberate
#                          probe of the exchangeability axis. The library
#                          requests dummies in multiples of p, so dim must
#                          divide p (the default dim = 3 would throw at
#                          p = 1000, hence dim = 5).
#
#  Scenario:
#  High-dimensional (n = 300, p = 1000, s = 10).
#  SNR sweep: {0.1, 0.2, 0.5, 0.6, 1.0, 2.0, 5.0}  (7 values, mirrors the
#  current C++ grid and its committed results).
#  Two support scenarios:
#    random support — redrawn per trial;
#    block support  — contiguous block {0, 1, ..., s-1}  (0-based).
#  Reports averaged FDR / TPR / Avg L / Avg T per distribution x SNR.
#
# Python downscale: num_MC = 10 (C++ runs 200) to keep the 3-solver x
# 12-distribution sweep moderate.
#
# ==============================================================================

import sys
import os

_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
_PARENT_DIR = os.path.dirname(_THIS_DIR)
for _p in (_THIS_DIR, _PARENT_DIR):
    if _p not in sys.path:
        sys.path.insert(0, _p)

import numpy as np

from trex_sim_common import run_mc_trex, save_mc_results

# ==============================================================================
# Global parameters
# ==============================================================================

_OUT_DIR     = os.path.join(_THIS_DIR, "simulation_results", "data")
_NUM_WORKERS = 6
_NUM_MC      = 10

# ==============================================================================
# Dummy distributions to compare
# ==============================================================================
#
# name : row label in output tables and CSV.
# spec : plain (picklable) dict resolved to a ts.DummyDistribution inside each
#        worker by trex_sim_common._make_dummy_distribution().

def _dist(name, dist_type, **params):
    return dict(name=name, spec=dict(type=dist_type, **params))

DISTRIBUTIONS = [
    _dist("Normal",         "Normal"),
    _dist("Uniform",        "Uniform"),
    _dist("Rademacher",     "Rademacher"),
    _dist("StudentT_df3",   "StudentT", df=3.0),
    _dist("StudentT_df5",   "StudentT", df=5.0),
    _dist("Laplace",        "Laplace"),
    _dist("Gumbel",         "Gumbel"),
    _dist("Triangle",       "Triangle"),
    _dist("Logistic",       "Logistic"),
    _dist("Mammen",         "Mammen"),
    _dist("SparseRad_s0.1", "ConstrainedSparseRademacher", s=0.1),
    _dist("UnifSphere_d5",  "UniformSphere", dim=5),
]

# Base solvers for the outer sweep (rho_afs only meaningful for TAFS)
_SOLVERS = [
    dict(solver_name="TLARS", name="TLARS", lambda2=0.1, rho_afs=0.3, ncgmp_variant=0),
    dict(solver_name="TOMP",  name="TOMP",  lambda2=0.1, rho_afs=0.3, ncgmp_variant=0),
    dict(solver_name="TAFS",  name="TAFS",  lambda2=0.1, rho_afs=0.3, ncgmp_variant=0),
]


# ==============================================================================
# MC simulation — dummy distribution comparison
# ==============================================================================

def trx_05_dummy_distributions(num_MC=_NUM_MC, num_workers=_NUM_WORKERS,
                               rnd_coef=False, block_support=False):

    n = 300
    p = 1000
    s = 10
    tFDR       = 0.1
    snr_values = [0.1, 0.2, 0.5, 0.6, 1.0, 2.0, 5.0]  # mirrors current C++ grid

    support_label = "block" if block_support else "random"
    print(f"\n{'=' * 70}")
    print("MC Simulation — Dummy Distribution Comparison  |  TLARS / TOMP / TAFS")
    print("  High-dimensional (p > n)")
    print(f"  n = {n},  p = {p},  s = {s},  num_MC = {num_MC}  [{support_label} support]\n")

    # Block support: contiguous 0-based {0, 1, ..., s-1}  (fixed across all trials)
    block_supp = np.arange(s, dtype=np.intp) if block_support else None

    n_dist = len(DISTRIBUTIONS)
    n_snr  = len(snr_values)
    dist_names = [d["name"] for d in DISTRIBUTIONS]

    # Solver sweep — one result file pair (txt + csv) per solver
    for solver_desc in _SOLVERS:

        # Result containers: distribution x SNR (reset per solver)
        fdr_mat   = np.zeros((n_dist, n_snr))
        tpr_mat   = np.zeros((n_dist, n_snr))
        avg_L_mat = np.zeros((n_dist, n_snr))
        avg_T_mat = np.zeros((n_dist, n_snr))

        # Simulation sweep: iterate over dummy distributions and SNR values
        for di, dist in enumerate(DISTRIBUTIONS):
            print(f"{'=' * 50}")
            print(f"Solver: {solver_desc['name']}  |  Dummy distribution: {dist['name']}")
            print(f"{'=' * 50}\n")

            # Per-distribution ctrl — only dummy_distribution varies
            base_ctrl = dict(
                K                        = 20,
                max_dummy_multiplier     = 10,
                use_max_T_stop           = True,
                lloop_strategy           = "STANDARD",
                dummy_distribution       = dist["spec"],
                parallel_rnd_experiments = False,   # mirrors R use_openmp=FALSE
            )

            for snr_idx, snr in enumerate(snr_values):
                base_seed = 24 + snr_idx * 1000
                lbl = f"SNR={snr:.2f} [{solver_desc['name']}/{dist['name']}]"
                result = run_mc_trex(
                    n=n, p=p, s=s, snr=snr,
                    solver_desc=solver_desc,
                    base_ctrl_dict=base_ctrl,
                    tFDR=tFDR,
                    base_seed=base_seed,
                    num_MC=num_MC,
                    max_workers=num_workers,
                    fixed_support=block_supp,   # None → random per trial
                    amplitude=1.0,
                    rnd_coef=rnd_coef,
                    track_L_T=True,
                    label=lbl,
                )
                fdr_mat[di, snr_idx]   = result["mean_FDR"]
                tpr_mat[di, snr_idx]   = result["mean_TPR"]
                avg_L_mat[di, snr_idx] = result["mean_L"]
                avg_T_mat[di, snr_idx] = result["mean_T"]
            print()

        stem = (f"demo_trex_05_dummy_distributions_results_n{n}_p{p}"
                f"_{solver_desc['name'].lower()}_{support_label}_support")
        save_mc_results(
            _OUT_DIR, stem, num_MC,
            dist_names, snr_values,
            fdr_mat, tpr_mat, avg_L_mat, avg_T_mat,
        )
        print()


# ==============================================================================
# Main
# ==============================================================================

if __name__ == "__main__":
    trx_05_dummy_distributions(block_support=False)
    # trx_05_dummy_distributions(block_support=True)
    print("\ndemo_trex_05_mc_sim_dummy_distributions.py complete.")
